Userland/Libraries/LibJS/Runtime/Value.cpp

```diff
@@ -474,6 +474,63 @@ FLATTEN ThrowCompletionOr<Value> Value::to_numeric(GlobalObject& global_object)
     return primitive.to_number(global_object);
 }
 
+struct NumberParseResult {
+    StringView literal;
+    u8 base;
+    bool invalid { false };
+};
+
+static Optional<NumberParseResult> parse_number_text(StringView text)
+{
+    NumberParseResult result {};
+
+    auto check_prefix = [&](auto lower_prefix, auto upper_prefix) {
+        if (text.length() <= 2)
+            return false;
+        if (!text.starts_with(lower_prefix) && !text.starts_with(upper_prefix))
+            return false;
+        return true;
+    };
+
+    if (check_prefix("0b"sv, "0B"sv)) {
+        if (!all_of(text.substring_view(2), is_ascii_binary_digit)) {
+            result.invalid = true;
+        } else {
+            result.literal = text.substring_view(2);
+            result.base = 2;
+        }
+    } else if (check_prefix("0o"sv, "0O"sv)) {
+        if (!all_of(text.substring_view(2), is_ascii_octal_digit)) {
+            result.invalid = true;
+        } else {
+            result.literal = text.substring_view(2);
+            result.base = 8;
+        }
+    } else {
+        bool is_signed = false;
+        if (text.starts_with('-')) {
+            text = text.substring_view(1);
+            is_signed = true;
+        } else if (text.starts_with('+')) {
+            text = text.substring_view(1);
+            is_signed = true;
+        }
+
+        if (check_prefix("0x"sv, "0X"sv)) {
+            if (is_signed || !all_of(text.substring_view(2), is_ascii_hex_digit)) {
+                result.invalid = true;
+            } else {
+                result.literal = text.substring_view(2);
+                result.base = 16;
+            }
+        } else {
+            return {};
+        }
```

Comment:
Why are you checking if it starts with a +/- and then only using that to make it invalid?
Can we not just let `strtod` handle the invalid strings starting with `0x`?
This is more explicit but it feels a little strange that this only happens for hexadecimal.

Reply:
`strtod` can parse hexadecimal numbers with signage, so I had to check for that specifically otherwise it would pass the invalid input to `strtod` and give out a number when it should be NaN. I do feel like it could be handled better, but I'm unsure of how I would go about it in a different way.

Reply:
I think `strtod` is inadvisable here, see #14691.